src/future/shared.rs

```diff
@@ -38,64 +39,76 @@ impl<F> fmt::Debug for Shared<F>
 {
     fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
         fmt.debug_struct("Shared")
-            .field("id", &self.id)
             .field("inner", &self.inner)
             .finish()
     }
 }
 
 struct Inner<F: Future> {
-    next_clone_id: Mutex<u64>,
-    state: Mutex<State<F>>,
+    next_clone_id: AtomicUsize,
+    future: UnsafeCell<Spawn<F>>,
+    result: UnsafeCell<Option<Result<SharedItem<F::Item>, SharedError<F::Error>>>>,
+    unparker: Arc<Unparker>,
 }
 
-impl<F> fmt::Debug for Inner<F>
-    where F: Future + fmt::Debug,
-          F::Item: fmt::Debug,
-          F::Error: fmt::Debug,
-{
-    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
-        fmt.debug_struct("Inner")
-            .field("next_clone_id", &self.next_clone_id)
-            .field("state", &self.state)
-            .finish()
-    }
+struct Unparker {
+    state: AtomicUsize,
+    waiters: Mutex<HashMap<usize, Task>>,
 }
 
-#[derive(Debug)]
-enum State<F: Future> {
-    Waiting(Arc<Unparker>, F),
-    Polling(Arc<Unparker>, Vec<task::Task>),
-    Done(Result<Arc<F::Item>, Arc<F::Error>>),
-}
+const IDLE: usize = 0;
+const POLLING: usize = 1;
+const REPOLL: usize = 2;
+const COMPLETE: usize = 3;
+const POISONED: usize = 4;
 
-impl<F> Shared<F>
-    where F: Future
-{
+impl<F> Shared<F> where F: Future {
     /// Creates a new `Shared` from another future.
     pub fn new(future: F) -> Self {
```

Comment:
Could you add a TODO and a `#[doc(hidden)]` here while you're at it? This wasn't ever supposed to be public...